src/tests/unit/common/simulated-browser-adapter.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { BrowserAdapter } from 'common/browser-adapters/browser-adapter';
+import { IMock, It, Mock } from 'typemoq';
+
+// This is a mock BrowserAdapter that maintains simulated state about which "windows" and "tabs"
+// exist and which listeners have been registered, and provides a few helper functions to simulate
+// certain browser events.
+export type SimulatedBrowserAdapter = IMock<BrowserAdapter> & {
+    // Tests may modify this state directly; updates will be reflected in the following default mock implementations:
+    //   * this.object.getAllWindows
+    //   * this.object.getTab
+    //   * this.object.tabsQuery
+    //
+    // Tests are responsible for maintaining self-consistency (ie, ensuring all tabs have corresponding windows)
+    windows: chrome.windows.Window[];
+    tabs: chrome.tabs.Tab[];
+
+    // These are set directly to whichever listener was last registered in the corresponding this.object.addListener* call
+    notifyOnConnect?: (port: chrome.runtime.Port) => void;
+    notifyTabsOnActivated?: (activeInfo: chrome.tabs.TabActiveInfo) => void;
+    notifyTabsOnUpdated?: (tabId: number, changeInfo: chrome.tabs.TabChangeInfo, tab: chrome.tabs.Tab) => void;
+    notifyTabsOnRemoved?: (tabId: number, removeInfo: chrome.tabs.TabRemoveInfo) => void;
+    notifyWebNavigationUpdated?: (details: chrome.webNavigation.WebNavigationFramedCallbackDetails) => void;
+    notifyWindowsFocusChanged?: (windowId: number) => void;
+
+    // These simulate real "update"/"activate" events (they update the windows/tabs state and send the notifications)
+    updateTab: (tabId: number, changeInfo: chrome.tabs.TabChangeInfo) => void;
+    activateTab: (tab: chrome.tabs.Tab) => void;
+};
+
+export function createSimulatedBrowserAdapter(tabs: chrome.tabs.Tab[], windows: chrome.windows.Window[]): SimulatedBrowserAdapter {
+    const mock: Partial<SimulatedBrowserAdapter> = Mock.ofType<BrowserAdapter>();
+    mock.tabs = [...tabs];
+    mock.windows = [...windows];
+    mock.setup(m => m.addListenerOnConnect(It.isAny())).callback(c => (mock.notifyOnConnect = c));
```

Comment:
should we use `It.is(isFunction)` here instead (`isFunction` can be imported from `lodash`)